theano/scan_module/scan_op.py

```diff
@@ -551,14 +551,36 @@ def make_thunk(self, node, storage_map, compute_map, no_recycling):
         # make_thunk can be called many times on the same op
         # we do not want to recompile the inner fct every time.
         if not getattr(self, 'fn', None):
-            self.fn = function(wrapped_inputs,
+
+            # Add a new input 'iter' to the inner function that will represent
+            # the index of the current iteration of the inner function
+            iter = theano.tensor.lscalar()
+            iter.tag.test_value = numpy.array(0, dtype="int64")
+            inputs = list(self.inputs)
+            givens = {}
+
+            # For every sequence element input to the inner function, replace
+            # the sequence element with a subtensor operation on the original
+            # sequence and add a new givens to perform the subtensor operation
+            # in the inner function
+            outer_seqs = self.outer_seqs(node)
+            for i in range(self.n_seqs):
+                inputs[i] = outer_seqs[i].type()
+                if hasattr(outer_seqs[i].tag, 'test_value'):
+                    inputs[i].tag.test_value = outer_seqs[i].tag.test_value
+
+                givens[self.inputs[i]] = inputs[i][iter]
+
+            self.fn = function([Param(p, borrow=True) for p in inputs+[iter]],
                                wrapped_outputs,
                                mode=self.mode_instance,
                                name=self.name,
+                               givens=givens,
                                profile=profile,
                                on_unused_input='ignore')
 
         try:
+            raise ImportError("Exception to force usage of the python version")
```

Comment:
Don't forget to remove this line and adapt the Cython implementation before the merge.